### PR TITLE
Change/6587 abstract ejb save

### LINKED ISSUE #6587: Create abstract base class for all infra. EJBs
### Feature Description
Sormas Infrastructure/Location

### Problem Description
We have a LOT of code duplication in the different EJBs which makes fixing and introducing new features a pain.

### Proposed Change
Create an abstract class for all EJBs and implement shared functionality there.

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #6587 
This PR cleans up `save` and `getByUuid`. Further, the EJBs now pass their criteria as generic. This is not important now, but will be necessary for a future generic implementation of count (which I have ready).

In review, we still need to clarify the following questions:

- [ ] Should every EJb really overwrite save and pass the validation error as argument? Should we use a cunstructor argument for it or create a function `onDuplicateError` or similar which is overvritten instead as we already do for `checkInfrastructureLock`. I really  would like to have better logging the case a duplicate was found, therefore, onDuplicateError could make more sense b/c then the logger could access getters of the concrete type to provide more info.
- [ ] Should find duplicate be part of `BaseFacade`? If yes should we do it later or in this PR?

### CODE REVIEW

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/country/CountryFacadeEjb.java [hunk 1]
```diff
@@ -198,37 +193,36 @@ public long count(CountryCriteria criteria) {
...
-			Optional<Country> byIsoCode = service.getByIsoCode(dto.getIsoCode(), true);
-			Optional<Country> byUnoCode = service.getByUnoCode(dto.getUnoCode(), true);
+			Optional<Country> byIsoCode = service.getByIsoCode(dtoToSave.getIsoCode(), true);
+			Optional<Country> byUnoCode = service.getByUnoCode(dtoToSave.getUnoCode(), true);
 			if (byIsoCode.isPresent() || byUnoCode.isPresent()) {
 				if (allowMerge) {
-					country = byIsoCode.isPresent() ? byIsoCode.get() : byUnoCode.get();
-					CountryDto dtoToMerge = getCountryByUuid(country.getUuid());
-					dto = DtoHelper.copyDtoValues(dtoToMerge, dto, true);
+					// todo no merging?
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/country/CountryFacadeEjb.java:210]: Can this be clarified and fixed? Or at least documented in a ticket so it doesn't go unnoticed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/district/DistrictFacadeEjb.java [hunk 2]
```diff
@@ -170,8 +165,8 @@ public DistrictDto getByUuid(String uuid) {
 			cq.where(filter);
 		}
 
-		if (sortProperties != null && sortProperties.size() > 0) {
-			List<Order> order = new ArrayList<Order>(sortProperties.size());
+		if (sortProperties != null && !sortProperties.isEmpty()) {
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/district/DistrictFacadeEjb.java:168]:
```suggestion
		if (CollectionUtils.isNotEmpty(sortProperties)) {
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/facility/FacilityFacadeEjb.java [hunk 3]
```diff
@@ -461,8 +458,8 @@ private FacilityDto toDto(Facility entity) {
 			cq.where(filter);
 		}
 
-		if (sortProperties != null && sortProperties.size() > 0) {
-			List<Order> order = new ArrayList<Order>(sortProperties.size());
+		if (sortProperties != null && !sortProperties.isEmpty()) {
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/facility/FacilityFacadeEjb.java:461]:
```suggestion
		if (CollectionUtils.isNotEmpty(sortProperties)) {
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/pointofentry/PointOfEntryFacadeEjb.java [hunk 4]
```diff
@@ -271,8 +243,8 @@ public void validate(PointOfEntryDto pointOfEntry) throws ValidationRuntimeExcep
 			cq.where(filter);
 		}
 
-		if (sortProperties != null && sortProperties.size() > 0) {
-			List<Order> order = new ArrayList<Order>(sortProperties.size());
+		if (sortProperties != null && !sortProperties.isEmpty()) {
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/pointofentry/PointOfEntryFacadeEjb.java:246]:
```suggestion
		if (CollectionUtils.isNotEmpty(sortProperties)) {
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/region/RegionFacadeEjb.java [hunk 5]
```diff
@@ -181,7 +179,7 @@ private void selectDtoFields(CriteriaQuery<RegionDto> cq, Root<Region> root) {
 			cq.where(filter);
 		}
 
-		if (sortProperties != null && sortProperties.size() > 0) {
+		if (sortProperties != null && !sortProperties.isEmpty()) {
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/region/RegionFacadeEjb.java:182]:
```suggestion
		if (CollectionUtils.isNotEmpty(sortProperties)) {
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/subcontinent/SubcontinentFacadeEjb.java [hunk 6]
```diff
@@ -159,7 +153,7 @@ public SubcontinentDto getByUuid(String uuid) {
 			cq.where(filter);
 		}
 
-		if (sortProperties != null && sortProperties.size() > 0) {
+		if (sortProperties != null && !sortProperties.isEmpty()) {
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/subcontinent/SubcontinentFacadeEjb.java:156]:
```suggestion
		if (CollectionUtils.isNotEmpty(sortProperties)) {
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/util/DtoHelper.java [hunk 7]
```diff
@@ -145,6 +145,7 @@ public static void fillDto(EntityDto dto, AbstractDomainObject entity) {
 		return target;
 	}
 
+	// todo this really should return void. Taking a target/receiver argument and then return it anyways does not make sense
```
Comment (vidi42) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/util/DtoHelper.java:148]: Will this be refactored with a different ticket?

**Review by vidi42 — CHANGES_REQUESTED** [review]

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/AbstractInfrastructureEjb.java [hunk 8]
```diff
@@ -23,6 +27,26 @@ protected AbstractInfrastructureEjb(SRV service, FeatureConfigurationFacadeEjb f
 		this.featureConfiguration = featureConfiguration;
 	}
 
+	protected DTO save(DTO dtoToSave, boolean allowMerge, String duplicateErrorMessage) {
```
[the remark this answers is not in the document]
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/AbstractInfrastructureEjb.java:30]: renamed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/country/CountryFacadeEjb.java (hunk 1 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/infrastructure/country/CountryFacadeEjb.java:210]: Follow up issue:  #6880

File: sormas-backend/src/main/java/de/symeda/sormas/backend/util/DtoHelper.java (hunk 7 above)
Comment (JonasCir) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/util/DtoHelper.java:148]: tracked in #6880

**Review by vidi42 — APPROVED** [review]

**Pull request merged**